File: data-structures/python/src/binary_search_tree.py
from typing import TypeVar, Generic, List, Iterator, Optional

T = TypeVar('T')


class BinarySearchTree(Generic[T]):
    class Node:
        def __init__(self, value: T) -> None:
            self.value: T = value
            self.left: Optional['BinarySearchTree.Node'] = None
            self.right: Optional['BinarySearchTree.Node'] = None

    def __init__(self) -> None:
        self._root: Optional[BinarySearchTree.Node] = None
        self._size: int = 0

    def insert(self, value: T) -> None:
        if self._root is None:
            self._root = BinarySearchTree.Node(value)
            self._size += 1
            return

        node = self._root
        while True:
            if value < node.value:
                if node.left is None:
                    node.left = BinarySearchTree.Node(value)
                    self._size += 1
                    return
                node = node.left
            elif value > node.value:
                if node.right is None:
                    node.right = BinarySearchTree.Node(value)
                    self._size += 1
                    return
                node = node.right
            else:
                return

    def remove(self, value: T) -> None:
        parent: Optional[BinarySearchTree.Node] = None
        node = self._root
        is_left_child = False

        while node is not None and node.value != value:
            parent = node
            if value < node.value:
                node = node.left
                is_left_child = True
            else:
                node = node.right
                is_left_child = False

        if node is None:
            return

        if node.left is None:
            replacement = node.right
        elif node.right is None:
            replacement = node.left
        else:
            successor_parent = node
            successor = node.right
            while successor.left is not None:
                successor_parent = successor
                successor = successor.left
            node.value = successor.value
            if successor_parent == node:
                successor_parent.right = successor.right
            else:
                successor_parent.left = successor.right
            self._size -= 1
            return

        if parent is None:
            self._root = replacement
        elif is_left_child:
            parent.left = replacement
        else:
            parent.right = replacement
        self._size -= 1

    def contains(self, value: T) -> bool:
        return self._find_node(self._root, value) is not None

    def min(self) -> T:
        if self._root is None:
            raise ValueError("min from empty tree")
        return self._find_min(self._root).value

    def max(self) -> T:
        if self._root is None:
            raise ValueError("max from empty tree")
        return self._find_max(self._root).value

    def size(self) -> int:
        return self._size

    def is_empty(self) -> bool:
        return self._size == 0

    def clear(self) -> None:
        self._root = None
        self._size = 0

    def in_order(self) -> List[T]:
        result: List[T] = []
        stack: List[BinarySearchTree.Node] = []
        node = self._root
        while stack or node is not None:
            while node is not None:
                stack.append(node)
                node = node.left
            node = stack.pop()
            result.append(node.value)
            node = node.right
        return result

    def pre_order(self) -> List[T]:
        result: List[T] = []
        if self._root is None:
            return result
        stack: List[BinarySearchTree.Node] = [self._root]
        while stack:
            node = stack.pop()
            result.append(node.value)
            if node.right is not None:
                stack.append(node.right)
            if node.left is not None:
                stack.append(node.left)
        return result

    def post_order(self) -> List[T]:
        result: List[T] = []
        if self._root is None:
            return result
        stack: List[BinarySearchTree.Node] = [self._root]
        while stack:
            node = stack.pop()
            result.append(node.value)
            if node.left is not None:
                stack.append(node.left)
            if node.right is not None:
                stack.append(node.right)
        result.reverse()
        return result

    def copy(self) -> 'BinarySearchTree[T]':
        clone: BinarySearchTree[T] = BinarySearchTree()
        for value in self.pre_order():
            clone.insert(value)
        return clone

    def _find_node(self, node: Optional[Node], value: T) -> Optional[Node]:
        while node is not None:
            if value < node.value:
                node = node.left
            elif value > node.value:
                node = node.right
            else:
                return node
        return None

    def _find_min(self, node: Node) -> Node:
        while node.left is not None:
            node = node.left
        return node

    def _find_max(self, node: Node) -> Node:
        while node.right is not None:
            node = node.right
        return node

    def __len__(self) -> int:
        return self._size

    def __contains__(self, value: T) -> bool:
        return self.contains(value)

    def __iter__(self) -> Iterator[T]:
        return iter(self.in_order())

    def __repr__(self) -> str:
        return f"BinarySearchTree({self.in_order()})"

    def __str__(self) -> str:
        return f"BinarySearchTree(size={self._size})"

File: data-structures/go/src/bst/binary_search_tree.go
package bst

import (
	"cmp"
	"errors"
)

var ErrEmptyTree = errors.New("BinarySearchTree: tree is empty")

type node[T cmp.Ordered] struct {
	value T
	left  *node[T]
	right *node[T]
}

type BinarySearchTree[T cmp.Ordered] struct {
	root *node[T]
	size int
}

func New[T cmp.Ordered]() *BinarySearchTree[T] {
	return &BinarySearchTree[T]{}
}

func (t *BinarySearchTree[T]) Insert(value T) {
	t.root = insertNode(t.root, value, &t.size)
}

func (t *BinarySearchTree[T]) Remove(value T) {
	t.root = removeNode(t.root, value, &t.size)
}

func (t *BinarySearchTree[T]) Contains(value T) bool {
	return findNode(t.root, value) != nil
}

func (t *BinarySearchTree[T]) Min() (T, error) {
	var zero T
	if t.root == nil {
		return zero, ErrEmptyTree
	}
	return findMin(t.root).value, nil
}

func (t *BinarySearchTree[T]) Max() (T, error) {
	var zero T
	if t.root == nil {
		return zero, ErrEmptyTree
	}
	return findMax(t.root).value, nil
}

func (t *BinarySearchTree[T]) Size() int {
	return t.size
}

func (t *BinarySearchTree[T]) IsEmpty() bool {
	return t.size == 0
}

func (t *BinarySearchTree[T]) Clear() {
	t.root = nil
	t.size = 0
}

func (t *BinarySearchTree[T]) InOrder() []T {
	result := make([]T, 0, t.size)
	inOrderTraverse(t.root, &result)
	return result
}

func (t *BinarySearchTree[T]) PreOrder() []T {
	result := make([]T, 0, t.size)
	preOrderTraverse(t.root, &result)
	return result
}

func (t *BinarySearchTree[T]) PostOrder() []T {
	result := make([]T, 0, t.size)
	postOrderTraverse(t.root, &result)
	return result
}

func (t *BinarySearchTree[T]) Clone() *BinarySearchTree[T] {
	clone := New[T]()
	cloneTree(t.root, clone)
	return clone
}

func insertNode[T cmp.Ordered](n *node[T], value T, size *int) *node[T] {
	if n == nil {
		*size++
		return &node[T]{value: value}
	}
	if value < n.value {
		n.left = insertNode(n.left, value, size)
	} else if value > n.value {
		n.right = insertNode(n.right, value, size)
	}
	return n
}

func removeNode[T cmp.Ordered](n *node[T], value T, size *int) *node[T] {
	if n == nil {
		return nil
	}

	if value < n.value {
		n.left = removeNode(n.left, value, size)
	} else if value > n.value {
		n.right = removeNode(n.right, value, size)
	} else {
		if n.left == nil {
			*size--
			return n.right
		}
		if n.right == nil {
			*size--
			return n.left
		}
		successor := findMin(n.right)
		n.value = successor.value
		n.right = removeNode(n.right, successor.value, size)
	}
	return n
}

func findNode[T cmp.Ordered](n *node[T], value T) *node[T] {
	for n != nil {
		if value < n.value {
			n = n.left
		} else if value > n.value {
			n = n.right
		} else {
			return n
		}
	}
	return nil
}

func findMin[T cmp.Ordered](n *node[T]) *node[T] {
	for n.left != nil {
		n = n.left
	}
	return n
}

func findMax[T cmp.Ordered](n *node[T]) *node[T] {
	for n.right != nil {
		n = n.right
	}
	return n
}

func inOrderTraverse[T cmp.Ordered](n *node[T], result *[]T) {
	if n == nil {
		return
	}
	inOrderTraverse(n.left, result)
	*result = append(*result, n.value)
	inOrderTraverse(n.right, result)
}

func preOrderTraverse[T cmp.Ordered](n *node[T], result *[]T) {
	if n == nil {
		return
	}
	*result = append(*result, n.value)
	preOrderTraverse(n.left, result)
	preOrderTraverse(n.right, result)
}

func postOrderTraverse[T cmp.Ordered](n *node[T], result *[]T) {
	if n == nil {
		return
	}
	postOrderTraverse(n.left, result)
	postOrderTraverse(n.right, result)
	*result = append(*result, n.value)
}

func cloneTree[T cmp.Ordered](n *node[T], clone *BinarySearchTree[T]) {
	if n == nil {
		return
	}
	clone.Insert(n.value)
	cloneTree(n.left, clone)
	cloneTree(n.right, clone)
}
